test/pthread/test_pthread_memory_growth.c

```diff
@@ -4,42 +4,64 @@
 // found in the LICENSE file.
 
 #include <pthread.h>
-#include <sys/types.h>
-#include <stdio.h>
 #include <stdlib.h>
 #include <assert.h>
-#include <emscripten.h>
+#include <emscripten/em_js.h>
+#include <stdbool.h>
+
+// We want to test that JavaScript access in the main thread automatically updates the memory views if the heap has grown from the pthread meanwhile.
+//
+// Checking this correctly is somewhat tricky because a lot of Emscripten APIs access heap on their own, so the memory views might be updated
+// by them before we explicitly check the state in our own JS routines below, which leads to false positives (test passing even though in
+// isolation the heap access in our own JS is not the one updating the memory views).
+//
+// To test it in isolation, we need to use only EM_JS - which is as close to a pure JS call as we can get - and not EM_ASM, pthread_join, proxying
+// etc., not even `puts`/`printf` for logging - as all of those access the heap from JS on their own and might update the memory views too early.
+// Not using standard proxying mechanisms also means we need to drop down all the way to raw atomics.
+
+EM_JS(void, assert_initial_heap_state, (bool isWorker), {
+  console.log(`Checking initial heap state on the ${isWorker ? 'worker' : 'main'} thread`);
```

Comment:
Can you use `dbg` here .. `console.log` is not the best thing on node workers and `dbg` will go direct to stdout.

Reply:
See https://github.com/emscripten-core/emscripten/issues/14804

Reply:
Oof TIL, that might explain some odd console.log behaviour I saw in the past. 

> I tested and this remains broken in node 14 but is fixed in node 15+. 

Sounds like it should be fixed now? Either way, yeah, happy to change.